include/seqan3/alphabet/composition.hpp

```diff
@@ -0,0 +1,211 @@
+ * This a "pure base class", you cannot instantiate it, you can only inherit from it.
+ * Most likely you are interested in using one of it's descendents like quality_composition.
+ *
+ * \sa quality_composition
+ * \sa mask_composition
+ */
+
+template <typename derived_type,
+          typename first_alphabet_type,
+          typename ...alphabet_types>
+      requires alphabet_concept<first_alphabet_type> && (alphabet_concept<alphabet_types> && ...)
+struct alphabet_composition :
+    public pod_tuple<first_alphabet_type, alphabet_types...>
+{
+public:
+    //!\brief The type of value_size and `alphabet_size_v<alphabet_composition<...>>`
+    using integral_type = detail::min_viable_uint_t<alphabet_size_v<first_alphabet_type> *
+                                                    (alphabet_size_v<alphabet_types> * ...)>;
+
+    //!\brief The product of the sizes of the individual alphabets.
+    static constexpr integral_type value_size{alphabet_size_v<first_alphabet_type> *
+                                              (alphabet_size_v<alphabet_types> * ...)};
+
+    //!\cond DEV
+    //!\brief the cummulative alphabet size products (first, first*second, first*second*third...) are cached
+    static constexpr std::array<integral_type, sizeof...(alphabet_types)+1> _cummulative_alph_sizes
+    {
+        [] () constexpr
+        {
+            std::array<integral_type, sizeof...(alphabet_types)+1> ret{};
+            size_t count = 0;
+            meta::for_each(meta::list<first_alphabet_type, alphabet_types...>{}, [&] (auto && alph) constexpr
+            {
+                ret[count] = alphabet_size_v<std::decay_t<decltype(alph)>> * (count > 0 ? ret[count - 1] : 1);
+                ++count;
+            });
+
+            return std::move(ret);
+        }()
+    };
+
+    //!\brief An index sequence up to the number of contained letters.
+    static constexpr auto _positions = std::make_index_sequence<sizeof...(alphabet_types)>{};
+    //!\endcond
+
+
+    /*!\name Read functions
+     * \{
+     */
+    /*!\brief Return the letter combinations numeric value or rank in the alphabet composition.
+     * \par Complexity
+     * Linear in the number of alpahabets.
+     */
+    constexpr integral_type to_integral() const
+    {
+        return to_integral_impl(_positions);
+    }
+    //!\}
+
+    /*!\name Write functions
+     * \{
+     */
+    /*!\brief Assign from a numeric value.
+     * \par Complexity
+     * Linear in the number of alpahabets.
+     * \par Exceptions
+     * Asserts that the parameter is smaller than value_size [only in debug mode].
+     */
+    constexpr derived_type & from_integral(integral_type const i)
+    {
+        assert(i < value_size);
+        from_integral_impl<0>(i);
+        return static_cast<derived_type &>(*this);
+    }
+    //!\}
+
+    //\brief Assign to the first letter that matches in the composition.
+    //TODO this is not passed on to descendents, why?
+//     template <typename type>
```

Comment:
Even if I have  an `assign` member, I still need to define an `operator=()` in both the base and the derived class to call if the `operator=()` is not inherited. So it would result in just more code and no added flexibility I think.